### ISSUE TITLE
UI and Buisness code separation

### ISSUE THREAD

**Issue by Doomsdayrs** [ATLauncher/ATLauncher#606]
**Is your feature request related to a problem? Please describe.**

UI and Business code is too tightly integrated. This leads to issues in which one must dig through the UI code to access business code too.

We should move to a more state based design.

**Describe the solution you'd like**

A slow migration from no architecture to basic architecture.

**Additional context**

We all can understand that having the launcher stuck in a state of messy UI code is no good.

**Comment by Doomsdayrs** [comment]
I will work on this 1 UI section at a time

**Comment by Doomsdayrs** [comment]
Personally from Android development, I enjoy MVVM patterns, as it is pretty fleshed out, and should provide the most improvement to the applications capabilities & maintainability.

Its also what I am most familiar with capable of producing.

**Comment by s0cks** [comment]
@Doomsdayrs 

I'm already working on refactoring the event system so it's decoupled w/ #604

I also have made a simple task engine for parallelizing the launcher, right now it can load the accounts, instances, & servers tab in the background. Going to create a MR for this soon.

My next ideal goals are to:
* Add Guice or some other DI framework, so creation of objects is a lot more straightforward.
* Add AspectJ or similar, to allow for reusable advice to remove repetative things like:
```java
PerformanceManager.start();
PerformanceManager.stop();
```

and allow more intelligent reactive programming to reduce event throughput like:

```java
@Subscribe
@OnSide(Side.UI)
public void onLocalizationChanged(final LocalizationChangedEvent event){
   // do something on the Swing thread
}
```

or

```java
@Subscribe
@OnSide(Side.ASYNC)
public void onLocalizationChanged(final LocalizationChangedEvent event){
  // do something on a worker thread
}
```

**Comment by s0cks** [comment]
@Doomsdayrs 
    
Yes please check the AccountManager, AccountsTab and related classes on my fork. These changes drastically separate the mvc.

 Now the app is reactive to when an account is added, removed or changed, and this can be done in the background as well, allowing the launcher to move on with the startup process and loading the accounts in the background updating as it loads them.

This is done with a combination of Guavas event bus, a work stealing executor service and some clever architecture of the event models

**Comment by s0cks** [comment]
Well it does deal with the UI as well, since most of what I'm building out is framework for the UI to be more parallel and decoupled

Essentially:
* hooking into the DI framework for construction of the components
* hooking into the task engine for async work
* Hooking into the event bus for reacting to "events"

These, can be used for all sorts of things such as:

** Updating the theme of the component when the theme is changed or uploaded 
** Updating the locale of the component when the locale is changes or uploaded
** Adding a ServerCard to the component when the app loads it in the background
** Searching for a mod pack in the background and rendering the results when the work completes on the UI

**Comment by s0cks** [comment]
> I can see what you mean by making it easier to make my changes, but vice versa, your PR is a large monolith of changes. Have you considered making smaller additions, one at a time instead? So it is easier to revert or debug if any issues arise?

It's not necessarily a large monolith of changes. It's one targeted feature. There are a lot of changes because this impacts a lot of things. This can always be squashed before merging in to reduce back-out overhead.

**Comment by Doomsdayrs** [comment]
@RyanTheAllmighty Where should view models and their interfaces be stored?

1. `gui/viewmodel/`
2. `viewmodel/`

**Comment by Doomsdayrs** [comment]
Alright, so aside from settings tab needing another rewrite, let us talk about packs tab.

The current situation is a nightmare, with UI & Business code tightly integrated far beyond any other part of the program.
I would have to rework it from the ground up, and I currently do not have the time to do so. Maybe during winter break.

**Comment by Doomsdayrs** [comment]
Vanilla packs was the second hardest, the first hardest will be a redesign of the packs tab. In which I have to analyze the code structure from ground up to properly create a compliant architecture replica.

I'll do settings tab before hand in one go, since I already have worked on it previously and I know what needs to be done.
